### PR TITLE
HPCC-23265 Add Unprotect All to WsDFU, etc

### PR DESCRIPTION
**Pull request by wangkx**
The Unprotect All allows an Admin user to remove the Protect flags
set by all users.

The WsDFU.DFUArrayAction is upgraded to support File Protection
and File Restriction.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 1]
```diff
@@ -1543,6 +1549,99 @@ bool CWsDfuEx::onDFUArrayAction(IEspContext &context, IEspDFUArrayActionRequest
...
+#endif
+
+    if (isDetachedFromDali())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
+
+    StringBuffer userID;
+    context.getUserID(userID);
+    Owned<IUserDescriptor> userDesc;
+    if (userID.isEmpty())
+        userID.set("hpcc"); //The userID is needed for setProtect().
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: @wangkx under what circumstances does it get here?  Only when running with no security?

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 2]
```diff
@@ -1543,6 +1549,99 @@ bool CWsDfuEx::onDFUArrayAction(IEspContext &context, IEspDFUArrayActionRequest
...
+        userDesc->set(userID, context.queryPassword(), context.querySignature());
+    }
+
+    for (unsigned i = 0; i < req.getLogicalFiles().length();i++)
+    {
+        const char *file = req.getLogicalFiles().item(i);
+        if (isEmptyString(file))
+            continue;
+
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(file, userDesc, false, false, true, nullptr, defaultPrivilegedUser); // lock super-owners
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Seems like you are changing some in the list and then if one isn't found you quit.  Should you gather exceptions and keep trying the rest of the list?

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 3]
```diff
@@ -1543,6 +1549,99 @@ bool CWsDfuEx::onDFUArrayAction(IEspContext &context, IEspDFUArrayActionRequest
...
+
+    for (unsigned i = 0; i < req.getLogicalFiles().length(); i++)
+    {
+        const char *file = req.getLogicalFiles().item(i);
+        if (isEmptyString(file))
+            continue;
+
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(file, userDesc, false, false, true, nullptr, defaultPrivilegedUser); // lock super-owners
+        if (!df)
+            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST, "Cannot find file %s.", file);
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Seems like you are changing some in the list and then if one isn't found you quit. Should you gather exceptions and keep trying the rest of the list?

**Review by afishbeck — COMMENTED** [review]
@wangkx a few questions / comments.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_dfu/ws_dfuService.cpp]: Yes, only when running with no security.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 2 above)
Comment (wangkx) [reply esp/services/ws_dfu/ws_dfuService.cpp]: Good point. Will do.

**Comment by wangkx** [comment]
@afishbeck please review my changes

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Pull request merged**